### PR TITLE
Circstats issue3705

### PR DESCRIPTION
**Pull request by mirca**
Hi,

Some circular statistical functions were implemented, as requested in issue #3705.

These functions are primarily based on those implemented in the R package [`CircStats`](https://cran.r-project.org/web/packages/CircStats/CircStats.pdf). Therefore, in most of the tests, results are compared with those obtained through similar functions presented in the aforementioned package.

Additionally, a function to compute the Cramér-Rao Lower Bounds for the von Mises distribution (which basically gives the uncertainties on estimating the parameters) was implemented. For this function, unfortunately, I could not find examples elsewhere in order to write a reliable enough test. Nonetheless, since the math envolved is simple and it only relies on the maximum likelihood estimation (which has been tested), I think this function is good.

I would be grateful for any comments.

### CODE REVIEW

**Comment by crawfordsm** [comment]
Thank for another pul request.    Some of these functions are already in scipy and I do not think we want to duplicate the work here unless we make the functions specifically useful for astronomy/astropy.  So I think the inclusion of these statistics does require some thoughtful discussion.  

One option would be that you should submit this pull request to scipy, which would make available to the large community and probably a good idea, as scipy already as a circstats module.  

However, a useful thing though would be making sure the stat functions work with `astropy.qunaitity` and that might be a good reason for their inclusion here.   In addition, you have added some more functionality which is not available in the scipy.stats.

**Comment by crawfordsm** [comment]
Also, are you considering implementing more of the circularly functionality from R?

**Comment by mirca** [comment]
Thanks @crawfordsm for commenting.

In fact, the `circmean` function is pretty much the same as the scipy one. However, the `circvar` function is not quite a duplication of the scipy one since they use a different approach to define "dispersion" for circular data.

I will definitely see what is necessary to the functions so that they work properly with `astropy.quantity`.

Yes. Do you have in mind any other functionality?

File: astropy/stats/circstats.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+    assert type(p) is int
+    if w is None:
+        w = np.ones(data.shape)
+    else:
+        assert w.shape == data.shape
+
+    C = np.sum(w*np.cos(p*(data-phi)), axis)/np.sum(w, axis)
+    S = np.sum(w*np.sin(p*(data-phi)), axis)/np.sum(w, axis)
+
+    return C,S
```
Comment (pllim) [inline astropy/stats/circstats.py]: Space after comma.

File: astropy/stats/circstats.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    # of the circular data.
+    data = np.asarray(data)
+
+    assert type(p) is int
+    if w is None:
+        w = np.ones(data.shape)
+    else:
+        assert w.shape == data.shape
+
+    C = np.sum(w*np.cos(p*(data-phi)), axis)/np.sum(w, axis)
```
Comment (pllim) [inline astropy/stats/circstats.py]: Spaces between variables and operators.

File: astropy/stats/circstats.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+__doctest_requires__ = {'vtest': ['scipy.stats'],
+                        'vonmisescrlb': ['scipy.special']}
+
+
+def _components(data, w=None, p=1, phi=0.0, axis=None):
+    # Utility function for computing the generalized rectangular components
+    # of the circular data.
+    data = np.asarray(data)
+
+    assert type(p) is int
```
Comment (pllim) [inline astropy/stats/circstats.py]: Is there a reason why `p` and only `p` is checked for the correct type? Since this is a hidden function, is this check really necessary?

File: astropy/stats/circstats.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+"""
+This module contains simple functions for dealing with circular statistics, for
+instance, mean, variance, standard deviation, correlation coefficient, and so
+on. This module also cover tests of uniformity, e.g the Rayleigh and V0 tests.
```
Comment (pllim) [inline astropy/stats/circstats.py]: "e.g.,"

File: astropy/stats/circstats.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+
+    return C,S
+
+def _angle(data, w=None, p=1, phi=0.0, axis=None):
+    # Utility function for computing the generalized sample mean angle
+    C,S = _components(data,w,p,phi,axis)
+    theta = np.arctan2(S,C)
+
+    mask = (S == .0)*(C == .0)
+    if mask.ndim > 0:
```
Comment (pllim) [inline astropy/stats/circstats.py]: Is the check needed? If `mask` is all `False`, the equation that follows will simply not assign any NaNs, right?

File: astropy/stats/circstats.py [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+       Series on Multivariate Analysis, Vol. 5, 2001.
+    .. [2] C. Agostinelli, U. Lund. "Circular Statistics from 'Topics in
+       Circular Statistics (2001)'". 2015.
+       <https://cran.r-project.org/web/packages/CircStats/CircStats.pdf>
+    """
+
+    alpha = np.asarray(alpha)
+    beta = np.asarray(beta)
+
+    assert np.size(alpha, axis=ax_alpha) == np.size(beta, axis=ax_beta)
```
Comment (pllim) [inline astropy/stats/circstats.py]: I think raising `ValueError` with a more descriptive error message will be more useful to the users.

File: astropy/stats/circstats.py [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+    rho = np.sum(sin_a*sin_b)/np.sqrt(np.sum(sin_a*sin_a)*np.sum(sin_b*sin_b))
+
+    return rho
+
+def rayleightest(data, w=None, axis=None):
+    """ Performs the Rayleigh test of uniformity.
+
+    This test is  used to indentify a non-uniform distribution, i.e. it is
+    designed for detecting an unimodal deviation from uniformity. More
+    precisely, it assumes the following hypotheses:
```
Comment (pllim) [inline astropy/stats/circstats.py]: I think Sphinx will complain if there is no newline between normal paragraph and bulleted list.

**Comment by pllim** [comment]
Just some nitpicking:

* Minor PEP8 formatting stuff.
* Perhaps it is better to replace `AssertionError` with `ValueError` and more descriptive error messages.
* The documentation is very nice but all buried within API docstring. They should be moved to a proper RST file as part of `stats` documentation.

File: astropy/stats/circstats.py (hunk 5 above)
Comment (mirca) [inline astropy/stats/circstats.py]: That's correct. In case `S == .0` and `C == .0` then  `np.arctan2(S,C)` returns `0`. However, I think `np.arctan2(0,0)` equals `np.nan` makes more sense (it is also the standard in R Package CircStats). I'm replacing by `if np.any(mask)` instead.

File: astropy/stats/circstats.py (hunk 3 above)
Comment (mirca) [inline astropy/stats/circstats.py]: Thanks. Actually, this is a misunderstanding. `p` may be real-valued.

File: astropy/stats/circstats.py (hunk 6 above)
Comment (embray) [inline astropy/stats/circstats.py]: I agree with @pllim; I generally don't like bare asserts outside of testing; if nothing else there should also be a message, but better to perform the check in an if statement and raise an explicit exception, if the check is even necessary at all.

File: astropy/stats/circstats.py [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+    >>> circmean(data) # doctest: +FLOAT_CMP
+    0.84870441244501904
+
+    References
+    ----------
+    .. [1] S. R. Jammalamadaka, A. SenGupta. "Topics in Circular Statistics".
+       Series on Multivariate Analysis, Vol. 5, 2001.
+    .. [2] C. Agostinelli, U. Lund. "Circular Statistics from 'Topics in
+       Circular Statistics (2001)'". 2015.
+       <https://cran.r-project.org/web/packages/CircStats/CircStats.pdf>
```
Comment (embray) [inline astropy/stats/circstats.py]: I don't see where this reference is referenced.

File: astropy/stats/circstats.py [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+
+    data = np.asarray(data)
+    n = np.size(data,axis=axis)
+    Rbar = _length(data,w,1,0.0,axis)
+    z = n*Rbar*Rbar
+
+    tmp = 1.0
+    if(n < 50):
+        tmp = 1.0 + (2.0*z - z*z)/(4.0*n) - (24.0*z - 132.0*z**2.0 +
+                76.0*z**3.0 - 9.0*z**4.0)/(288.0*n*n)
```
Comment (embray) [inline astropy/stats/circstats.py]: I understand these are probably some fitted parameters, but these numbers all deserve explanation or at least some reference.

Comment (mirca) [inline astropy/stats/circstats.py]: @embray Thanks, I've added a reference.

File: astropy/stats/circstats.py (hunk 8 above)
Comment (mirca) [inline astropy/stats/circstats.py]: @pllim Thanks, replaced the `asserts` by `ValueError`. Also, it seems that all functions in the stats package have their documentation within docstring.

**Comment by crawfordsm** [comment]
@astronomeralex, @eteq, and @mhvk -- I'm curious about your thoughts about inclusion of the code:

pros: It is code useful for pulsar astronomy as mentioned in #3705 and has several functions that are not currently available elsewhere.

cons: some functionality overlaps with [scipy.stats](http://docs.scipy.org/doc/scipy/reference/stats.html#circular-statistical-functions) and it is more generally statistical methods rather than astronomy specific.

File: astropy/stats/circstats.py (hunk 8 above)
Comment (pllim) [inline astropy/stats/circstats.py]: @mirca, yes, you are right. In this case, the doc you have is consistent with the package. Never mind. :)

**Comment by astronomeralex** [comment]
@crawfordsm @eteq @mhvk That's a good question. I would lean towards including the code for your pro's listed above. This, and other pull requests, keep pressing this issue of whether stats are astro-specific. It seems prudent that we solve this question for the whole module instead of pull by pull.

**Comment by mhvk** [comment]
@crawfordsm - I agree with you: these routines are useful outside of astronomy as well and the aim should be to have them in scipy. The main objection I can see is that for some it would be useful to have them available *now* rather than wait for things to have moved to a stable release in scipy. For that, we could have our versions in a `utils.compat.scipy` module, in analogy with what we currently do for numpy (see https://github.com/astropy/astropy/pull/2327).

In the process, we would hopefully also resolve any issues with scipy not being compatible with `Quantity` (and therefore likely having problems with other `ndarray` subclasses), which would benefit everyone as well. Indeed, @mirca, if you adapt your tests to work with `Quantity`, I'd love to have those as part of tests for `astropy.units`! 

@mirca - is this something you are willing to pursue?  At some level, you'd be our guinea pig, to see how such PRs fare.

p.s.  For the same reasons as described above, I do think it would be good to eventually move `astropy.units` to `scipy` as well. My main hesitation so far has been that, without `__numpy_ufunc__`, the state of `Quantity` is not yet as ideal as it could be.

File: astropy/stats/circstats.py [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+
+    References
+    ----------
+    .. [1] S. R. Jammalamadaka, A. SenGupta. "Topics in Circular Statistics".
+       Series on Multivariate Analysis, Vol. 5, 2001.
+    .. [2] C. Agostinelli, U. Lund. "Circular Statistics from 'Topics in
+       Circular Statistics (2001)'". 2015.
+       <https://cran.r-project.org/web/packages/CircStats/CircStats.pdf>
+    .. [3] M. Chirstman., C. Miller. "Testing a Sample of Directions for
+       Uniformity." Lecture Notes, STA 6934/5805. University of Florida, 2007.
```
Comment (embray) [inline astropy/stats/circstats.py]: Not that this isn't a useful reference as-is.  I was able to Google for it, and find it helpful.  But is there not a reference for this that is less ephemeral than some lecture notes, which have a habit of disappearing?  

Basically, if there are magic numbers appearing in some code, those numbers have been be well-substantiated in the code itself, or at least reference an easy to find book or something, because otherwise the code is at high risk of becoming a black box.  Maybe I'm over-thinking it in this case though.  I don't know how common this derivation is.

Comment (mirca) [inline astropy/stats/circstats.py]: @mhvk How this adaptation would look like? You mean using the input data as `Quantity` objects?

**Comment by mhvk** [comment]
@mirca - yes, one would like to be sure the code works with `Quantity`, including if the quantity is in degrees. Since the first thing the present code does is turn any input into a straight `ndarray`, this is unlikely to work as is. Note that I think that here one should just ducktype (or at least do `np.asanyarray`); most of the arithmatic should just work with quantities.

**Comment by mhvk** [comment]
OK, I did what I should have done before commenting too much. First, I tried and the `scipy` routine definitely is not `Quantity` safe. Second, I looked at the [scipy code](https://github.com/scipy/scipy/blob/v0.16.1/scipy/stats/morestats.py#L2352), and think it will be somewhat problematic to make this work with `Quantity`, mostly because they use `high` and `low` values that have defaults that are float rather than of some unit, and are clearly meant to do the unit conversion that is automatic if one uses Quantity. I worry though that they would be less than keen to start inserting that are quantity-specific.

Given the above and given that the code they use is relatively trivial and even easier to write using `Quantity`, it may just makes sense to do that after all.

That said, I do think any code in astropy should fully use `Quantity`. I'm also wary of using definitions that are different from those used in `scipy` (even if they are consistent with what is used in `R`).

Just to show how the code might look like if full Quantity-aware
```
def circmean(sample, axis=None):
    sample = Longitude(sample)
    with u.add_enabled_equivalencies(u.dimensionless_angles()):
        complex_mean = np.exp(1j * sample).mean(axis=axis)
    return Longitude(np.arctan2(complex_mean.imag, complex_mean.real),
                     sample.unit, wrap_angle=sample.wrap_angle)

from astropy.coordinates import Longitude
theta = Longitude(np.random.uniform(np.pi,-np.pi,10), u.rad)
circmean(theta)
# <Longitude 4.086819352329046 rad>
circmean(theta.to(u.deg))
# <Longitude 234.15750052084297 deg>
theta.to(u.deg)
# <Longitude [ 318.40807173, 258.09501963, 205.94616268, 256.10717963,
#               86.79577277,  19.49205343, 186.6603956 , 231.79219607,
#              109.07587581, 249.56135721] deg>
```

Alternatively, one could wrap the `scipy` code where possible:
```
def circmean(sample, axis=None):
    sample = Longitude(sample)
    circmean = scipy.stats.circmean(sample.to(u.rad.value), axis=axis)
    return Longitude(circmean * u.rad, sample.unit, wrap_angle=sample.wrap_angle)
```

**Comment by crawfordsm** [comment]
@mirca  Once you have updated the code to work with Quantities, I'll give it a final review before merging.  Although it looks like you will need to rebase most likely due to updates to the change log.

**Comment by mirca** [comment]
Hi @crawfordsm, @mhvk. I've made some modifications and now the functions are able to handle both ``Quantity`` and non-Quantity types. More precisely, if the input ``data`` is a ``Quantity`` object, then the returned value, when appropriate, will be also a ``Quantity`` object. On the other hand, if the input ``data`` is not a ``Quantity`` object, then the returned value will not be a ``Quantity`` either. Is that the desired behavior? 
I also add a parameter called ``deg``, in which the user should set to ``True`` whenever the circular data is expressed in degrees.

**Comment by mhvk** [comment]
@mirca - had only the most cursory of looks so far, and liked very much that for most of the code there is, as I would have hoped, absolutely no difference between what is done for `ndarray` and for `Quantity`. This is duck-typing at its best! 

The exceptions are the degree to radian conversions; my sense would be to not allow this. Effectively, this is equivalent to allowing some special units to be attached to `ndarray`, which is just what `Quantity` is for. So, while I'm quite happy for plain numbers in an `ndarray` to be considered to have dimension of radian (as this is equivalent to dimensionless, which is what a plain number should be), I don't like the idea of allowing plain numbers to really be degrees.

**Comment by mirca** [comment]
@crawfordsm, @mhvk, Does this look better? (:

I made some modifications in such a way that all functions now will return ``Quantity`` regardless if the input user is a ``ndarray`` or a ``Quantity`` object. That basically removes the need to convert deg to rad.

**Comment by crawfordsm** [comment]
@mirca  I will try to review it this week and give you comments.

**Comment by mirca** [comment]
@crawfordsm any thoughts?

File: astropy/stats/circstats.py [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+        weighting factor for each group such that sum(w,axis) equals the number
+        of observations. See [1], remark 1.4, page 22, for detailed
+        explanation.
+    axis : int, optional
+        Axis along which circular variances are computed. The default is to
+        compute the variance of the flattened array.
+
+    Returns
+    -------
+    circvar : float
```
Comment (crawfordsm) [inline astropy/stats/circstats.py]: returns quantity

File: astropy/stats/circstats.py [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+        weighting factor for each group such that sum(w,axis) equals the number
+        of observations. See [1], remark 1.4, page 22, for detailed
+        explanation.
+    axis : int, optional
+        Axis along which circular means are computed. The default is to compute
+        the mean of the flattened array.
+
+    Returns
+    -------
+    circmean : float
```
Comment (crawfordsm) [inline astropy/stats/circstats.py]: returns quanity

**Comment by crawfordsm** [comment]
Sorry to take so long to get back to you on the comments.  Thanks for the high quality pull request and the work you have done on it.   Overall the functionality and the tests seem to be work.   I've added some minor inline comments and it would be good to take care of the minor pep8 issues in the test failures (I think the other ones are unrelated to this pull request).   Also great job with the tests and having tests against the R-package results.

The only thing I don't really like is the `deg` option.   I think this should be removed as it was pretty non-intuitive and I actually made the mistake right off the bat of trying to pass it a degree quantity without knowing that should change.

Since this is an astropy function, my preferred choice is that only quantities should be passed to it so there isn't any confusion about the unit of it.   However, if you do want to still be able to pass `numpy.ndarrays` to it, then it should either always assume those are in radians.  I think the third option of ignoring the degree value when a quantity is passed to it will also lead to some confusion on how the task works and should be avoided.

File: astropy/stats/tests/test_circstats.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+    beta_deg = beta_deg*u.deg
+    assert_allclose(answer, circcorrcoef(alpha_deg, beta_deg, deg=True),
+                    atol=1e-4)
+    # testing input in radians
+    alpha_rad = alpha_rad*u.rad
+    beta_rad = beta_rad*u.rad
+    assert_allclose(answer, circcorrcoef(alpha_rad, beta_rad), atol=1e-4)
+
+
+def test_rayleightest1():
```
Comment (crawfordsm) [inline astropy/stats/tests/test_circstats.py]: Can you give this a more descriptive name than just test1?   Or at least a comment about what is different between test1 and test2 and what they are suppose to specifcially  test?

File: astropy/stats/circstats.py [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+    >>> circvar(data_deg*u.deg, deg=True) # doctest: +FLOAT_CMP
+    <Quantity 0.16356352748437508>
+
+    References
+    ----------
+    .. [1] S. R. Jammalamadaka, A. SenGupta. "Topics in Circular Statistics".
+       Series on Multivariate Analysis, Vol. 5, 2001.
+    .. [2] C. Agostinelli, U. Lund. "Circular Statistics from 'Topics in
+       Circular Statistics (2001)'". 2015.
+       <https://cran.r-project.org/web/packages/CircStats/CircStats.pdf>
```
Comment (crawfordsm) [inline astropy/stats/circstats.py]: It would be good to add a `Note` about the difference between this and the circvar in scipy

File: astropy/stats/circstats.py [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+    centered : Boolean, optinal
+        If ``True``, central circular moments are computed. Default value is
+        ``False``.
+    axis : int, optional
+        Axis along which circular moments are computed. The default is to
+        compute the circular moment of the flattened array.
+
+    Returns
+    -------
+    circmoment: np.ndarray
```
Comment (crawfordsm) [inline astropy/stats/circstats.py]: quantity

File: astropy/stats/circstats.py [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+        beta = Quantity(beta, u.rad)
+
+    mu_a = circmean(alpha, deg, w_alpha, ax_alpha)
+    mu_b = circmean(beta, deg, w_beta, ax_beta)
+
+    sin_a = np.sin(alpha - mu_a)
+    sin_b = np.sin(beta - mu_b)
+    rho = np.sum(sin_a*sin_b)/np.sqrt(np.sum(sin_a*sin_a)*np.sum(sin_b*sin_b))
+
+    return rho
```
Comment (crawfordsm) [inline astropy/stats/circstats.py]: Since this is dimensionalless, probably should just return the value rather than a quantity

Comment (mhvk) [inline astropy/stats/circstats.py]: I agree with @crawfordsm's general comment that the `deg` flag should be removed. It will make the code much cleaner; and fine *not* to convert to `Quantity` at all, so that `ndarray` just gets treated as radians. In that case, I also think it is fine for the output here to just be of the type of the input, i.e., dimensionless `Quantity` if `Quantity` input was used.

**Comment by mirca** [comment]
@crawfordsm @mhvk Thank you for the review!

**Comment by mirca** [comment]
Hi @crawfordsm, @mhvk,
sorry for taken a bit long to implement the requested changes.

The `deg` flag was removed as requested. Also, the coding still works with both ``Quantity`` and ``non-Quantity`` inputs. However, ``non-Quantity`` inputs are assumed to be in radians (is this the desired behavior?). Further, some functions returns dimensionless ``Quantity`` if a ``Quantity`` input was used.

File: astropy/stats/circstats.py [hunk 17]
```diff
@@ -0,0 +1,448 @@
...
+
+def _angle(data, w=None, p=1, phi=0.0, axis=None):
+    # Utility function for computing the generalized sample mean angle
+    C, S = _components(data, w, p, phi, axis)
+
+    # theta will be an angle in the interval [-np.pi, np.pi)
+    # [-180, 180)*u.deg in case data is a Quantity
+    theta = np.arctan2(S, C)
+
+    if isinstance(data, Quantity):
```
Comment (mhvk) [inline astropy/stats/circstats.py]: Why is this needed?

File: astropy/stats/circstats.py [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+    .. [1] S. R. Jammalamadaka, A. SenGupta. "Topics in Circular Statistics".
+       Series on Multivariate Analysis, Vol. 5, 2001.
+    .. [2] C. Agostinelli, U. Lund. "Circular Statistics from 'Topics in
+       Circular Statistics (2001)'". 2015.
+       <https://cran.r-project.org/web/packages/CircStats/CircStats.pdf>
+    """
+    phi = 0.0
+    if centered:
+        phi = circmean(data, w, axis)
+
```
Comment (mhvk) [inline astropy/stats/circstats.py]: Really nitpicky, but why not
```
else:
    phi = 0.0
```
(i.e., do not have the assignment upfront)

File: astropy/stats/circstats.py [hunk 19]
```diff
@@ -0,0 +1,448 @@
...
+__all__ = ['circmean', 'circvar', 'circmoment', 'circcorrcoef', 'rayleightest',
+           'vtest', 'vonmisesmle']
+__doctest_requires__ = {'vtest': ['scipy.stats']}
+
+
+def _components(data, w=None, p=1, phi=0.0, axis=None):
+    # Utility function for computing the generalized rectangular components
+    # of the circular data.
+    if w is None:
```
Comment (mhvk) [inline astropy/stats/circstats.py]: One could be slightly more general here, allowing the weights to be broadcasted:
```
if w is None:
    w = np.ones((1,))
try:
    w = broadcast_to(w, data.shape)
except ValueError:
    raise ValueError('Weights and data have inconsistent shape.')
```
(Note that you need to add `from astropy.utils.compat.numpy import broadcast_to` since it is only in recent numpy versions.)

File: astropy/stats/circstats.py (hunk 17 above)
Comment (mirca) [inline astropy/stats/circstats.py]: It seems that ``np.arctan2`` is always going to return ``u.rad`` even if the input is ``u.deg``. So, I used ``if isinstance(data, Quantity):`` (a bit afraid, I confess) in order to return ``u.deg`` when appropriate.

File: astropy/stats/circstats.py (hunk 18 above)
Comment (mirca) [inline astropy/stats/circstats.py]: Thanks, readability counts :)

File: astropy/stats/circstats.py [hunk 20]
```diff
@@ -0,0 +1,448 @@
...
+        w = np.ones(data.shape)
+    elif (w.shape != data.shape):
+        raise ValueError('shape of w %s is not equals shape of data %s.'
+                         % (w.shape, data.shape,))
+
+    C = np.sum(w * np.cos(p * (data - phi)), axis)/np.sum(w, axis)
```
Comment (mhvk) [inline astropy/stats/circstats.py]: To speed things up, I would do
```
angle = p * (data - phi)
C = np.average(np.cos(angle), weights=w, axis=axis)
S = np.average(np.sin(angle), weights=w, axis=axis)
```
This way you actually do not need to check `w` at all, and it will be faster for `w=None`, since `np.average` then skips the multiplication.

File: astropy/stats/circstats.py [hunk 21]
```diff
@@ -0,0 +1,448 @@
...
+    return theta
+
+
+def _length(data, w=None, p=1, phi=0.0, axis=None):
+    # Utility function for computing the generalized sample length
+    C, S = _components(data, w, p, phi, axis)
+    return np.hypot(S, C)
+
+
+def circmean(data, w=None, axis=None):
```
Comment (mhvk) [inline astropy/stats/circstats.py]: By analogy with `np.average`, should the order be `data, axis, weight`?

File: astropy/stats/circstats.py [hunk 22]
```diff
@@ -0,0 +1,448 @@
...
+    beta : numpy.ndarray
+        Array of circular (directional) data.
+    w_alpha : numpy.ndarray, optional
+        In case of grouped data, the i-th element of ``w_alpha`` represents a
+        weighting factor for each group such that sum(w,axis) equals the number
+        of observations. See [1], remark 1.4, page 22, for detailed
+        explanation.
+    w_beta : numpy.ndarray, optional
+        See description of ``w_alpha``.
+    ax_alpha : int, optional
```
Comment (mhvk) [inline astropy/stats/circstats.py]: Should we really do separate axes for the two arrays? Maybe it is cleaner to just give one axis and let the caller organise things? I wonder mostly since broadcasting along the other axes could make things somewhat surprising.

File: astropy/stats/circstats.py [hunk 23]
```diff
@@ -0,0 +1,448 @@
...
+
+    if w is None:
+        w = np.ones(data.shape)
+    elif (w.shape != data.shape):
+        raise ValueError('shape of w %s is not equals shape of data %s.'
+                         % (w.shape, data.shape,))
+
+    n = np.size(data, axis=axis)
+
+    R0bar = np.sum(w*np.cos(data-mu), axis)/np.sum(w, axis)
```
Comment (mhvk) [inline astropy/stats/circstats.py]: In analogy with above, this could be
```
R0bar = np.average(np.cos(data-mu), weights=w, axis=axis)
```
in which case the tests on `w` can be removed.

**Comment by mhvk** [comment]
@mirca -as you will have seen, I again made more comments, which I think make this even better. Sorry to be so late with these; the last month has been hectic and I haven't really kept up with astropy.

File: astropy/stats/circstats.py (hunk 17 above)
Comment (mhvk) [inline astropy/stats/circstats.py]: I would not worry about that: the user can change it to degrees if they want to; all we care for here is that it is an angle unit. (If you do think this is important then I would for any quantity do `theta = theta.to(data.unit)`, so that the user gets the same unit back; after all, it could be arcseconds!).

File: astropy/stats/circstats.py (hunk 23 above)
Comment (mirca) [inline astropy/stats/circstats.py]: It seems that ``np.average`` gets rid of ``Quantity``, i.e.,
`>>> a = np.array([30, 45, 60])*u.deg`
`>>> np.average(a)`
`45.0`

**Comment by mirca** [comment]
@mhvk Thanks a lot! (:

File: astropy/stats/circstats.py (hunk 23 above)
Comment (mhvk) [inline astropy/stats/circstats.py]: @mirca - well-caught! So, agreed to stick to your implementation.

**Comment by mirca** [comment]
@mhvk @crawfordsm Perhaps this is ready for final review?

File: astropy/stats/circstats.py [hunk 24]
```diff
@@ -0,0 +1,447 @@
...
+    C, S = _components(data, p, phi, axis, weights)
+    return np.hypot(S, C)
+
+
+def circmean(data, axis=None, weights=None):
+    """ Computes the circular mean angle of an array of circular data.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
```
Comment (crawfordsm) [inline astropy/stats/circstats.py]: This could be an numpy.ndarray or Quantity

File: astropy/stats/circstats.py [hunk 25]
```diff
@@ -0,0 +1,447 @@
...
+        Array of circular (directional) data.
```
Comment (crawfordsm) [inline astropy/stats/circstats.py]: Could you add a comment that if an numpy.ndarray it assumes it is in radians?

File: astropy/stats/circstats.py [hunk 26]
```diff
@@ -0,0 +1,447 @@
...
+        the mean of the flattened array.
+    weights : numpy.ndarray, optional
+        In case of grouped data, the i-th element of ``w`` represents a
+        weighting factor for each group such that sum(w,axis) equals the number
+        of observations. See [1], remark 1.4, page 22, for detailed
+        explanation.
+
+    Returns
+    -------
+    circmean : Quantity
```
Comment (crawfordsm) [inline astropy/stats/circstats.py]: If an numpy.ndarray, it returns a numpy.ndarray not a quanity which is okay, but this docstring needs to be updated

**Comment by crawfordsm** [comment]
Thanks for all the updates @mirca.  

A final note is that the docstrings need to be made consistent with how the functions work.   I've left an example for `circmean`, but could you update all functions where it is appropriate?  Other than that, I'll be happy to have this merged.

**Comment by mirca** [comment]
@crawfordsm Nice catch!

**Comment by crawfordsm** [comment]
Looks like it

**Comment by mirca** [comment]
@crawfordsm done! :)

**Comment by crawfordsm** [comment]
Thanks @mirca !
